File: Python For Everyone, Horstmann/Chapter5-Functions/P5.26.py
# Write a program that reads in a bar code (with  : denoting half bars and |  denoting
# full bars) and prints out the zip code it represents. Print an error message if the bar
# code is not correct.

# IMPORTS
from sys import exit

# FUNCTIONS
def barcodeToNum(barcode):
    if barcode == "||:::": return 0
    elif barcode == ":::||": return 1
    elif barcode == "::|:|": return 2
    elif barcode == "::||:": return 3
    elif barcode == ":|::|": return 4
    elif barcode == ":|:|:": return 5
    elif barcode == ":||::": return 6
    elif barcode == "|:::|": return 7
    elif barcode == "|::|:": return 8
    elif barcode == "|:|::": return 9

    else:
        exit("Unknown barcode")


def convert(barcode):
    NDIGIT = 6
    DIGIT_LENGTH = 5

    if len(barcode) != 2 + NDIGIT * DIGIT_LENGTH:
        exit("Invalid barcode")

    if barcode[0] != "|":
        exit("Invalid barcode starting character. Likely a wrong barcode too")

    if barcode[-1] != "|":
        exit("Invalid barcode ending character. Likely a wrong barcode too")

    d1 = barcodeToNum(barcode[1:DIGIT_LENGTH + 1])
    d2 = barcodeToNum(barcode[1 + DIGIT_LENGTH:2 * DIGIT_LENGTH + 1])
    d3 = barcodeToNum(barcode[1 + 2 * DIGIT_LENGTH:3 * DIGIT_LENGTH + 1])
    d4 = barcodeToNum(barcode[1 + 3 * DIGIT_LENGTH:4 * DIGIT_LENGTH + 1])
    d5 = barcodeToNum(barcode[1 + 4 * DIGIT_LENGTH:5 * DIGIT_LENGTH + 1])
    d6 = barcodeToNum(barcode[1 + 5 * DIGIT_LENGTH:6 * DIGIT_LENGTH + 1])

    if (d1 + d2 + d3 + d4 + d5 + d6) % 10 != 0:
        exit("Invalid barcode sum")

    return 10000 * d1 + 1000 * d2 + 100 * d3 + 10 * d4 + d5

# main
def main():
    print("Example input: ||:|:::|:|:||::::::||:|::|:::|||")
    print("Example expected output: 95014")

    inputBarcode = str(input("Enter a postal barcode: "))

    value = convert(inputBarcode)
    if value >= 0:
        print("The value is", value)

    else:
        print("The postal bar code is invalid")

# PROGRAM RUN
main()

File: Big-Java-for-Everyone-Late-Objects-Horstmann/Chapter-5-Methods/P5_26.java
/*Write a program that reads in a bar code (with : denoting half bars and | denoting
full bars) and prints out the zip code it represents. Print an error message if the bar
code is not correct.*/

import java.util.Scanner;

public class P5_26 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter a bar code: ");
        String barcode = input.next();
        input.close();
        if (barcode.length() != 32) {
            System.out.println("Barcode is invalid.");
        } else {
            int sum = 0;
            int zip = 0;
            for (int i = 0; i < 5; i++) {
                int digit = barcodeToZip(barcode.substring(i + i * 5, 7 + i * 5));
                if (digit == -1) {
                    System.out.println("Encountered an invalid barcode digit.");
                    return;
                }
                zip = digit + 10 * zip;
                sum += digit;
            }
            int checkDigit = barcodeToZip(barcode.substring(1 + 5 * 5, 7 + 5 * 5));
            if ((checkDigit + sum) % 10 != 0) {
                System.out.println("Error checkdigit mismatch.");
            } else {
                System.out.println(zip);
            }
        }
    }

    public static int barcodeToZip(String barCode) {
        if (barCode.equals("||:::")) {
            return 0;
        }
        int digit = 0;
        if (barCode.charAt(0) == '|') {
            digit += 7;
        }
        if (barCode.charAt(1) == '|') {
            digit += 4;
        }
        if (barCode.charAt(2) == '|') {
            digit += 2;
        }
        if (barCode.charAt(3) == '|') {
            digit += 1;
        }
        if ((digit == 1 || digit == 2 || digit == 4 || digit == 7) && barCode.charAt(4) != '|') {
            return -1;
        } else if (digit < 1 || digit > 9) {
            return -1;
        } else {
            return digit;
        }
    }
}
